cloud/pkg/operationcontroller/controller/downstream.go

```diff
@@ -0,0 +1,360 @@
+		unscheduleNode.Labels = map[string]string{NodeUpgradeJobStatusKey: NodeUpgradeJobStatusValue}
+		byteNode, err := json.Marshal(unscheduleNode)
+		if err != nil {
+			klog.Warningf("marshal data failed: %v", err)
+			continue
+		}
+
+		_, err = dc.kubeClient.CoreV1().Nodes().Patch(context.Background(), node, apimachineryType.StrategicMergePatchType, byteNode, metav1.PatchOptions{})
+		if err != nil {
+			klog.Errorf("failed to drain node %s: %v", node, err)
+			continue
+		}
+	}
+}
+
+func needUpgrade(node *v1.Node, upgradeVersion string) bool {
+	if filterVersion(node.Status.NodeInfo.KubeletVersion, upgradeVersion) {
+		klog.Warningf("Node(%s) version(%s) already on the expected version %s.", node.Name, node.Status.NodeInfo.KubeletVersion, upgradeVersion)
+		return false
+	}
+
+	// we only care about edge nodes, so just remove not edge nodes
+	if !isEdgeNode(node) {
+		klog.Warningf("Node(%s) is not edge node", node.Name)
+		return false
+	}
+
+	// if node is in Upgrading state, don't need upgrade
+	if _, ok := node.Labels[NodeUpgradeJobStatusKey]; ok {
+		klog.Warningf("Node(%s) is in upgrade state", node.Name)
+		return false
+	}
+
+	// if node is in NotReady state, don't need upgrade
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status != v1.ConditionTrue {
+			klog.Warningf("Node(%s) is in NotReady state", node.Name)
+			return false
+		}
+	}
+
+	return true
+}
+
+// handleNodeUpgradeJobTimeout is used to handle the situation that cloud don't receive upgrade result from edge node
+// within the timeout period
+func (dc *DownstreamController) handleNodeUpgradeJobTimeout(node string, upgradeID string, upgradeVersion string, historyID string, timeoutSeconds *uint32) {
+	// by default, if we don't receive upgrade response in 300s, we think it's timeout
+	// if we have specified the timeout in Upgrade, we'll use it as the timeout time
+	var timeout uint32 = 300
+	if timeoutSeconds != nil {
+		timeout = *timeoutSeconds
+	}
+
+	receiveFeedback := false
+
+	// check whether edgecore report to the cloud about the upgrade result
+	err := wait.Poll(10*time.Second, time.Duration(timeout)*time.Second, func() (bool, error) {
+		v, ok := dc.nodeUpgradeJobManager.UpgradeMap.Load(upgradeID)
+		if !ok {
+			// we think it's receiveFeedback to avoid construct timeout response by ourselves
+			receiveFeedback = true
+			klog.Errorf("NodeUpgradeJob %v not exist", upgradeID)
+			return false, fmt.Errorf("nodeUpgrade %v not exist", upgradeID)
+		}
+		upgradeValue := v.(*v1alpha1.NodeUpgradeJob)
+		for index := range upgradeValue.Status.UpgradeStatus {
+			if upgradeValue.Status.UpgradeStatus[index].NodeName == node {
+				// if HistoryID matches and state is v1alpha1.Completed
+				// it means we've received the specified Upgrade Operation
+				if upgradeValue.Status.UpgradeStatus[index].History.HistoryID == historyID &&
+					upgradeValue.Status.UpgradeStatus[index].UpgradeState == v1alpha1.Completed {
```

Comment:
here HistoryID is used to check this upgrade operation is successful or not, if not, we'll construct a timeout record.